Add docs for `start_spoof` and `stop_spoof` (#2123)
software-mansion/protostar: website/docs/cairo-1/06-testing/cheatcodes-reference/start_spoof.md

# `start_spoof`

```cairo
fn start_spoof(contract_address: felt252, mock: TxInfoMock) nopanic;
```

Changes `TxInfo` returned by `get_tx_info()` for the targeted contract until the spoof is stopped
with [stop_spoof](./stop_spoof.md).

- `contract_address` address of the contract for which `get_tx_info()` result will be mocked.

> - `TxInfoMock` - a struct with same structure as `TxInfo` (returned by `get_tx_info()`), fields are optional to allow
>   partial mocking of the `TxInfo`, i.e. for those fields that are `Option::None`, `get_tx_info` will return original
>   values.

Reviewer: ```suggestion
- `TxInfoMock` - a struct with same structure as `TxInfo` (returned by `get_tx_info()`), 

To mock the field of `TxInfo`, set the corresponding field of `TxInfoMock` to `Some(mocked_value)`. Setting the field to `None` will use a default value - the field will not be mocked. Using `None` will also cancel current mock for that field. See below for practical example.
```

struct TxInfoMock {
    version: Option<felt252>,
    account_contract_address: Option<felt252>,
    max_fee: Option<u128>,
    signature: Option<Array<felt252>>,
    transaction_hash: Option<felt252>,
    chain_id: Option<felt252>,
    nonce: Option<felt252>,
}

trait TxInfoMockTrait {
    // Returns a default object initialized with Option::None for each field  
    fn default() -> TxInfoMock;
}
```

```cairo title="Contract example"
#[contract]
mod SimpleContract {
    use box::BoxTrait;
    use starknet::info::get_tx_info;

    struct Storage {
        stored_hash: felt252
    }

    #[external]
    fn store_tx_hash() {
        let tx_info = get_tx_info().unbox();
        stored_hash::write(tx_info.transaction_hash);
    }

    #[view]
    fn get_stored_tx_hash() -> felt252 {
        stored_hash::read()
    }
}
```

```cairo title="Test example"
use array::ArrayTrait;
use result::ResultTrait;
use cheatcodes::TxInfoMockTrait;
use option::OptionTrait;

#[test]
fn test_start_spoof() {
    let contract_address = deploy_contract('simple', @ArrayTrait::new()).unwrap();
    let version_before_mock = call(contract_address, 'get_transaction_version', @ArrayTrait::new()).unwrap();

>     // Set tx_hash to 1234

Reviewer: ```suggestion
    // Change transaction_hash to 1234
    // All other fields of `TxInfo` remain unchanged
```

    let mut tx_info = TxInfoMockTrait::default();
    tx_info.transaction_hash = Option::Some(1234);
    start_spoof(contract_address, tx_info);

    // Stores tx_hash in contract storage
    invoke(contract_address, 'store_tx_hash', @ArrayTrait::new()).unwrap();
    // Retrieve stored tx_hash
    let return_data = call(contract_address, 'get_stored_tx_hash', @ArrayTrait::new()).unwrap();
    assert(*return_data.at(0_u32) == 1234, *return_data.at(0_u32));

>     // Verify that none of the other fields have been mocked

Reviewer: ```suggestion
    // Verify that only transaction_hash has been mocked
```

# `start_spoof`

```cairo
fn start_spoof(contract_address: felt252, mock: TxInfoMock) nopanic;
```

Changes `TxInfo` returned by `get_tx_info()` for the targeted contract until the spoof is stopped
with [stop_spoof](./stop_spoof.md).

- `contract_address` address of the contract for which `get_tx_info()` result will be mocked.
- `TxInfoMock` collection of fields equivalent to those in `TxInfo` (returned by `get_tx_info()`).

> All fields are optional to allow partial mocking of the `TxInfo`, i.e. for those fields that are `Option::None`, `get_tx_info` will return original values.

Reviewer: Lets just call it a struct since it is just that. 

Also I'd add a better explanation on how passing `Option::None` works in term of not mocking specific fields or cancelling the mocks.

```cairo title="TxInfoMock"
struct TxInfoMock {
    version: Option<felt252>,
    account_contract_address: Option<felt252>,
    max_fee: Option<u128>,
    signature: Option<Array<felt252>>,
    transaction_hash: Option<felt252>,
    chain_id: Option<felt252>,
    nonce: Option<felt252>,
}

trait TxInfoMockTrait {
    fn default() -> TxInfoMock;
}

impl TxInfoMockImpl of TxInfoMockTrait {
    fn default() -> TxInfoMock {
        TxInfoMock {
            version: Option::None(()),
            account_contract_address: Option::None(()),
            max_fee: Option::None(()),
            signature: Option::None(()),
            transaction_hash: Option::None(()),
            chain_id: Option::None(()),
            nonce: Option::None(()),
        }
    }

> }

Reviewer: Not sure if we need to show the implementation here. Perhaps just a comment in the trait would be enough?

```

```cairo title="Contract example"
#[contract]
mod SimpleContract {
    use box::BoxTrait;
    use starknet::info::get_tx_info;

    struct Storage {
        stored_hash: felt252
    }

    #[external]
    fn store_tx_hash() {
        let tx_info = get_tx_info().unbox();
        stored_hash::write(tx_info.transaction_hash);
    }

    #[view]
    fn get_stored_tx_hash() -> felt252 {
        stored_hash::read()
    }
}
```

```cairo title="Test example"
use array::ArrayTrait;
use result::ResultTrait;
use cheatcodes::TxInfoMockTrait;
use option::OptionTrait;

#[test]
fn test_start_spoof() {
    let contract_address = deploy_contract('simple', @ArrayTrait::new()).unwrap();
    let version_before_mock = call(contract_address, 'get_transaction_version', @ArrayTrait::new()).unwrap();

    let mut tx_info = TxInfoMockTrait::default();
    tx_info.transaction_hash = Option::Some(1234);
    start_spoof(contract_address, tx_info);

    invoke(contract_address, 'store_tx_hash', @ArrayTrait::new()).unwrap();
    let return_data = call(contract_address, 'get_stored_tx_hash', @ArrayTrait::new()).unwrap();

>     assert(*return_data.at(0_u32) == 1234, *return_data.at(0_u32));

Reviewer: I think this example could use some comments to explain what's happening.

>     // 

Reviewer 2: remove?
